docs/source/rfc-configuration-overhaul.md

```diff
@@ -0,0 +1,893 @@
+```
+
+While this message is more helful than the ones we discussed above, there are still issues:
+
+- The message contains repetition.
+- Without an input snippet, the `Invalid character 's' at position 4` part of the message is not as helpful as it could
+  be.
+
+#### 8. Config file with extra fields
+
+Providing a configuration file with extra fields that are not supposed to be there does not produce any errors at all.
+
+This might lead to a bad user experience when user expects some options to apply, but doesn't have any idea that those
+in fact are silently ignored.
+
+#### 9. Specifying `SUMERAGI.PUBLIC_KEY`
+
+As you can see in the "Redundant fields" section, user should not specify any non-null value for `SUMERAGI.PUBLIC_KEY`
+parameter. However, if they do, they will get the following error:
+
+```
+Error:
+   0: Please add `PUBLIC_KEY and PRIVATE_KEY` to the configuration.
+```
+
+The problem is that the error message tells something completely unrelated to the actual cause. This produces a terrible
+experience for users.
+
+#### 10. Config file is an invalid JSON file
+
+Providing an invalid JSON file as a config results in the same error we've seen before, which is not useful at all in
+this particular case and doesn't tell the user anything about the invalid JSON:
+
+```
+Error:
+   0: Please add `PUBLIC_KEY and PRIVATE_KEY` to the configuration.
+```
+
+### Ambiguity in Setting Root-Level Fields via ENV
+
+To understand the issue with ambiguity at the root-level fields, let's once again look at the excerpt from the
+[configuration reference](https://github.com/hyperledger/iroha/blob/35ba182e1d1b6b594712cf63bf448a2edefcf2cd/docs/source/references/config.md#network):
+
+> **`network`**
+>
+> Network configuration
+>
+> Has type `Option<network::ConfigurationProxy>`. Can be configured via environment variable `IROHA_NETWORK`
+>
+> ```json
+> {
+>   "ACTOR_CHANNEL_CAPACITY": 100
+> }
+> ```
+
+As we can see, `network` is not actually a configuration field in itself. Instead, there are various `network.*` fields
+that can be configured. This bring a question of what will happen when we set the `IROHA_NETWORK` environment variable
+as mentioned in the excerpt above? Will it override the nested `IROHA_NETWORK_CAPACITY` fields? Is there a use case for
+providing an ability to set `IROHA_NETWORK` all-in-one through an environment variable?
+
+### Chaotic Code Organisation
+
+Internally, not all of the configuration-related logic is contained within a single `iroha_config` crate. Instead, some
+of the configuration resolution logic is located in other crates, such as `iroha_cli`. This makes the
+configuration-related code error-prone and harder to maintain.
+
+## Proposals
+
+The proposals outlined in this section are designed to collectively enhance the configuration system within Iroha.
+However, they are not mutually required. Each proposal can be considered and implemented independently.
+
+### Proposal 1 - Use TOML
```

Comment:
how about `yaml` alternative proposal?
common for deployments

Reply:
Well, choice between TOML and YAML is more about a personal preference, I believe. However, we prefer TOML primarily because:

- Simple and explicit syntax, more readable
- Strong types, less unexpected results
- Indentation doesn't matter
- Better error messages
- It is a standard for Rust ecosystem

As for making YAML/JSON/whatever an additional configuration format, it might badly affect the overall quality of the configuration system due to it being more generalised, thus e.g. less able to provide some details in error messages.

Reply:
IMO `TOML` is the simplest of the three and is best suited for Iroha purposes

Reply:
My main motivation is that configuration is rather related to deployment, whereas `TOML` is common for a project (development) configuration in Rust, though I don't deny there deployment tools relying on `TOML` may exist. DevOps input may be really valuable here.

Reply:
While my current position is unrelated to DevOps, I'm confident I have enough experience to discuss DevOps issues.

For me, `TOML` is more readable, and the advantage of `JSON`/`YAML`/`TOML` is that we learn them quickly as human beings.

If we assume that people are configuring Iroha, they either know `TOML` or will know it as soon as they need to use it.
If we assume we're using GPT to configure it (for some reason), it knows the format and will most likely get stuck on the variable names. Real people may make the same mistake unless they have the config reference doc for the latest Iroha version with a lot of examples.

> whereas TOML is common for a project (development) configuration in Rust, though I don't deny there deployment tools relying on TOML may exist

TOML is intended to be a configuration file format, according to Tom Preston-Werner at https://toml.io/en/: 

> A config file format for humans.

> TOML aims to be a minimal configuration file format that's easy to read due to obvious semantics. TOML is designed to map unambiguously to a hash table. TOML should be easy to parse into data structures in a wide variety of languages.

Reply:
IMO `yaml` is unnecessarily complicated and it's a mess to deal with spaces. On the other hand, `JSON` would be fine except that it doesn't support comments, and we really want to be able to comment out parameters in the config file. `TOML` fits our needs perfectly

if you look at `mysql.conf` it's configuration format is quite similar to `TOML`.